core/modules/widgets/navigator.js

```diff
@@ -112,22 +107,43 @@ NavigatorWidget.prototype.replaceFirstTitleInStory = function(storyList,oldTitle
 		storyList.splice(0,0,newTitle);
 	}
 };
-
+	
 NavigatorWidget.prototype.addToStory = function(title,fromTitle) {
 	var storyList = this.getStoryList();
-	if(storyList) {
-		// See if the tiddler is already there
-		var slot = this.findTitleInStory(storyList,title,-1);
-		// If not we need to add it
-		if(slot === -1) {
-			// First we try to find the position of the story element we navigated from
-			slot = this.findTitleInStory(storyList,fromTitle,-1) + 1;
-			// Add the tiddler
-			storyList.splice(slot,0,title);
-			// Save the story
-			this.saveStoryList(storyList);
+	// Quit if we cannot get hold of the story list
+	if(!storyList) return;
+	// See if the tiddler is already there
+	var slot = storyList.indexOf(title);
+	// Quit if it already exists in the story river
+	if(slot >= 0) return;
+	// First we try to find the position of the story element we navigated from
+	var fromIndex = storyList.indexOf(fromTitle);
+	if(fromIndex >= 0) {
+		// How to open internal links that were clicked from *within* the story river?
+		var openLinkFromInsideRiver = $tw.wiki.getTiddlerText(
+			"$:/config/Navigation/openLinkFromInsideRiver", "below");
+		// The tiddler is added from inside the river
+		// Determine where to insert the tiddler; Fallback is "below"
+		switch(openLinkFromInsideRiver) {
+			case "top": slot = 0; break;
```

Comment:
Again, it's not specified explicitly, I'm afraid, but TW coding style would be to keep the statements of the 'case' clauses split across multiple lines.

```
	case "top":
		slot = 0;
		break;
```